### PR TITLE
Add support for promoteId

### PR DESCRIPTION
**Pull request by ank27**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Add support for PromoteId to be used with Feature state API.</changelog>`.

### Summary of changes

This PR adds support for `promoteId` which allows a developer to promote a particular string / source-property to be the feature state id in features.

'promoteId' is a feature for geojson and vector sources. The feature allows to promote feature's property to a feature id, so that promoted id can be used with FeatureState API.

### CODE REVIEW

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    fun fromProperty(propertyName: Any): PromoteId {
+      return when (propertyName) {
+        is String -> {
+          PromoteId(propertyName = propertyName)
+        }
+        is HashMap<*, *> -> {
+          @Suppress("UNCHECKED_CAST")
+          val propertyMap = propertyName as HashMap<String, String>
+          val key = propertyMap.keys.iterator().next()
+          PromoteId(key, propertyMap[key]!!)
```
Comment (Chaoba) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:42]: Can we avoid using `!!` here?

**Review by Chaoba — APPROVED** [review]
LGTM

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * Holds a property type to promote a specific feature for feature state API.
+ *
+ * @param sourceId source layer id of the feature, either source [GeoJsonSource] or [VectorSource].
+ * @param propertyName feature property name.
+ *
+ * For more information see [The online documentation]https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#promoteId).
+ */
+@Keep
+data class PromoteId @JvmOverloads constructor(
+  val sourceId: String? = null,
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:18]: if this param is optional - shouldn't it be actually second one? guess it will make usage easier

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Holds a property type to promote a specific feature for feature state API.
+ *
+ * @param sourceId source layer id of the feature, either source [GeoJsonSource] or [VectorSource].
+ * @param propertyName feature property name.
+ *
+ * For more information see [The online documentation]https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#promoteId).
+ */
+@Keep
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:16]: Are we using keep for other similar classes? just curious about it

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+     * Can be either of type [String] or [HashMap] of <String,String>
+     */
+    fun fromProperty(propertyName: Any): PromoteId {
+      return when (propertyName) {
+        is String -> {
+          PromoteId(propertyName = propertyName)
+        }
+        is HashMap<*, *> -> {
+          @Suppress("UNCHECKED_CAST")
+          val propertyMap = propertyName as HashMap<String, String>
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:40]: what if it's not `HashMap<String, String>`? Will we have runtime crash? Perhaps use try-catch here to produce better exception text

Comment (kiryldz) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:40]: I'm not sure how it works, could check like:
```
is HashMap<String, String> -> {
```
instead of 
```
is HashMap<*, *> -> {
```
work here?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    return Value(propertyName)
+  }
+
+  companion object {
+    /**
+     * Construct a [PromoteId] object from a Property returned from the core.
+     * Can be either of type [String] or [HashMap] of <String,String>
+     */
+    fun fromProperty(propertyName: Any): PromoteId {
+      return when (propertyName) {
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:34]:
```suggestion
    fun fromProperty(propertyName: Any) = when (propertyName) {
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * @param propertyName feature property name.
+ *
+ * For more information see [The online documentation]https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#promoteId).
+ */
+@Keep
+data class PromoteId @JvmOverloads constructor(
+  val sourceId: String? = null,
+  val propertyName: String
+) {
+  internal fun toValue(): Value {
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:21]: Is it used in tests only?
if yes - do we need this function at all here? we could add it as extension function in tests or do smth similar

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import com.mapbox.maps.extension.style.sources.generated.GeoJsonSource
+import com.mapbox.maps.extension.style.sources.generated.VectorSource
+
+/**
+ * Holds a property type to promote a specific feature for feature state API.
+ *
+ * @param sourceId source layer id of the feature, either source [GeoJsonSource] or [VectorSource].
+ * @param propertyName feature property name.
+ *
+ * For more information see [The online documentation]https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#promoteId).
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:14]:
```suggestion
 * For more information see [The online documentation](https://docs.mapbox.com/mapbox-gl-js/style-spec/types/#promoteId).
```
And also not sure this will work as expected in real generated docs

Comment (kiryldz) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:14]: Also link does not lead me to promote id...

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+import androidx.annotation.Keep
+import com.mapbox.bindgen.Value
+import com.mapbox.maps.extension.style.sources.generated.GeoJsonSource
+import com.mapbox.maps.extension.style.sources.generated.VectorSource
+
+/**
+ * Holds a property type to promote a specific feature for feature state API.
+ *
+ * @param sourceId source layer id of the feature, either source [GeoJsonSource] or [VectorSource].
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:11]: Do we have a check for that?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    fun fromProperty(propertyName: Any): PromoteId {
+      return when (propertyName) {
+        is String -> {
+          PromoteId(propertyName = propertyName)
+        }
+        is HashMap<*, *> -> {
+          @Suppress("UNCHECKED_CAST")
+          val propertyMap = propertyName as HashMap<String, String>
+          val key = propertyMap.keys.iterator().next()
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:41]: what if we pass an empty hash map? 😄

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    sourceId?.let {
+      return Value(hashMapOf(sourceId to Value(propertyName)))
+    }
+    return Value(propertyName)
+  }
+
+  companion object {
+    /**
+     * Construct a [PromoteId] object from a Property returned from the core.
+     * Can be either of type [String] or [HashMap] of <String,String>
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:31]: Specify `@throws` as well as `@param`

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 2 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:18]: i did it the same way initially, but then aligned it with the docs here 
https://docs.mapbox.com/mapbox-gl-js/style-spec/sources/#vector-promoteId. 
`Either a property name, or an object of the form {<sourceLayer>: <propertyName>}`

But definitely its better to use optional param as last args.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 3 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:16]: Yes we are. :D

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 4 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:40]: Using HashMap<String,String> gives compile time error.
`Cannot check for instance of erased type: kotlin.collections.HashMap<String, String> /* = java.util.HashMap<String, String>`

Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:40]: I will add more try{} catch{} for hashmap, but the args for this function returned from core in either string or map<string,string>.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 7 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:14]: Yeah, the link is broken from gl-js side, I have asked @alexshalamov if that will be fixed or i will remove from here :)

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 9 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:40]: fair point, will add a check here. :D

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 8 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:11]: `promoteId` is only available for GeoJson or vector source on [builder method](https://github.com/mapbox/mapbox-maps-android/pull/636/files#diff-f5e000d4ded37a7719c6f0253cb88496cf0820543ec87717a2eed34dfd169eadR586) so we don't need to check here.

But if you are asking that if we initialize promoteId with the wrong sourceId and provide that to geojson builder -> IMO it should be handled by core itself, and shouldn't be applied to feature state, but I will test it.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 6 above)
Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:21]: it also used in [Builder methods](https://github.com/mapbox/mapbox-maps-android/blob/fd16ddb63b0f065b178fea7fa3b681a925620c70/extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt#L586-L587) of source in order to add the properties to source.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    fun fromProperty(propertyName: Any) = when (propertyName) {
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:33]: Could you add tests for this method and cover all scenarios?

File: extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  fun fromProperty_UnsupportedDataClassType() {
+    PromoteId.fromProperty(MockType("source", "abc"))
+  }
+
+  @Test(expected = RuntimeException::class)
+  fun fromProperty_InvalidHashMapType() {
+    PromoteId.fromProperty(hashMapOf(1 to 2))
+  }
+
+  internal data class MockType(val a: String, val b: String)
```
Comment (alexander-kulikovskii) [inline extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt:53]:
```suggestion
  private data class MockType(val a: String, val b: String)
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            val key = propertyMap.keys.iterator().next()
+            PromoteId(propertyMap[key] ?: "", key)
+          } else {
+            PromoteId("")
+          }
+        } catch (e: RuntimeException) {
+          throw RuntimeException("$propertyName must be in the format HashMap<String,String>")
+        }
+      }
+      else -> throw RuntimeException("Wrapping ${propertyName::class.java.simpleName} to PromoteId is not supported.")
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:52]:
```suggestion
      else -> throw UnsupportedTypeException("Wrapping ${propertyName::class.java.simpleName} to PromoteId is not supported.")
```

File: extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt [hunk 14]
```diff
@@ -0,0 +1,54 @@
+package com.mapbox.maps.extension.style.utils
+
+import com.mapbox.maps.extension.style.types.PromoteId
+import org.junit.Assert.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
```
Comment (alexander-kulikovskii) [inline extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt:9]: do really need this runner?

File: extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Test
+  fun fromProperty_EmptyString() {
+    val expected = PromoteId("")
+    assertEquals(expected, PromoteId.fromProperty(""))
+  }
+
+  @Test
+  fun fromProperty_HashMap() {
+    val expected = PromoteId("abc", "source")
```
Comment (alexander-kulikovskii) [inline extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt:26]: nice to add one more test `PromoteId("abc", null)`

Comment (alexander-kulikovskii) [reply extension-style/src/test/java/com/mapbox/maps/extension/style/utils/PromoteIdTest.kt:26]: and `PromoteId("", "source")`

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+            val key = propertyMap.keys.iterator().next()
+            PromoteId(propertyMap[key] ?: "", key)
+          } else {
+            PromoteId("")
+          }
+        } catch (e: RuntimeException) {
+          throw RuntimeException("$propertyName must be in the format HashMap<String,String>")
+        }
+      }
+      else -> throw UnsupportedOperationException("Wrapping ${propertyName::class.java.simpleName} to PromoteId is not supported.")
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:53]: can we somehow check this exception?

**Review by kiryldz — APPROVED** [review]

**Review by alexander-kulikovskii — APPROVED** [review]

File: extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt (hunk 11 above)
Comment (pengdev) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:35]: I think we could split this function to two functions to make the API signature strongly typed, since we only support two types of data.  i.e. `fromProperty(propertyName: String)` and `fromProperty(propertyMap: Map<String, String>)`. 

Thinking again, why would we need to have this method? The promoteId could either a property name, or an object of the form {<sourceLayer>: <propertyName>}, so the constructor should be well enough for this purpose.

Comment (ank27) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/types/PromoteId.kt:35]: > I think we could split this function to two functions to make the API signature strongly typed, since we only support two types of data. i.e. `fromProperty(propertyName: String)` and `fromProperty(propertyMap: Map<String, String>)`.
> 
> Thinking again, why would we need to have this method? The promoteId could either a property name, or an object of the form {: }, so the constructor should be well enough for this purpose. cc @ank27

It was made this way to handle the errors in case core provides other type formats. 
hmm, I see that we can use `Value` as it is already wrapper for type.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/VectorSource.kt [hunk 17]
```diff
@@ -301,6 +321,16 @@ class VectorSource(builder: Builder) : Source(builder.sourceId) {
       properties[propertyValue.propertyName] = propertyValue
     }
 
+    /**
+     * A property to use as a feature id (for feature state). Either a property name, or
+     * an object of the form `{<sourceLayer>: <propertyName>}`. If specified as a string for a vector tile
+     * source, the same property is used across all its source layers.
+     */
+    fun promoteId(value: PromoteId) = apply {
+      val propertyValue = PropertyValue("promoteId", TypeUtils.wrapToValue(value.toValue()))
```
Comment (pengdev) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/VectorSource.kt:330]: Nit: do we still need wrapToValue since toValue() should have been Value type already?

**Review by pengdev — APPROVED** [review]
LGTM with a nit
